tensorflow/core/kernels/gather_functor_gpu.cu.h

```diff
@@ -66,11 +66,56 @@ __global__ void GatherOpKernel(const T* __restrict__ params,
       // out.
       Index params_i =
           (batch_i * gather_dim_size + gather_i) * slice_size + slice_i;
-      out[i] = ldg(params + params_i);
+      out[i] = params[params_i];
     }
   }
 }
 
+namespace detail {
+
+template <bool is_axis_zero>
+struct LaunchGatherKernelVectorized {
+  template <int vec_size>
+  struct Impl {
+    template <typename T, typename Index>
+    Status operator()(const GPUDevice& d, const T* params, const Index* indices,
+                      T* out, int64 gather_dim_size, int64 indices_size,
+                      int64 slice_size, int64 out_size) {
+      CHECK(slice_size % vec_size == 0);  // Crash OK
+      CHECK(out_size % vec_size == 0);    // Crash OK
+      CHECK(reinterpret_cast<std::uintptr_t>(params) % vec_size ==
+            0);                                                      // Crash OK
+      CHECK(reinterpret_cast<std::uintptr_t>(out) % vec_size == 0);  // Crash OK
+      int64 out_size_vec = out_size / vec_size;
+      int64 slice_size_vec = slice_size / vec_size;
+      using Tvec = AlignedVector<T, vec_size>;
+      const Tvec* params_vec = reinterpret_cast<const Tvec*>(params);
+      Tvec* out_vec = reinterpret_cast<Tvec*>(out);
+
+      GpuLaunchConfig config = GetGpuLaunchConfig(
+          out_size_vec, d, &GatherOpKernel<Tvec, Index, is_axis_zero>,
+          /*dynamic_shared_memory_size=*/0, /*block_size_limit=*/0);
+      return GpuLaunchKernel(
+          GatherOpKernel<Tvec, Index, is_axis_zero>, config.block_count,
+          config.thread_per_block, 0, d.stream(), params_vec, indices, out_vec,
+          gather_dim_size, indices_size, slice_size_vec, out_size_vec);
+    }
+  };
+};
+
+}  // namespace detail
+
+template <bool is_axis_zero, typename T, typename Index>
+Status LaunchGatherKernel(const GPUDevice& d, const T* params,
+                          const Index* indices, T* out, int64 gather_dim_size,
+                          int64 indices_size, int64 slice_size,
+                          int64 out_size) {
+  return DispatchToVectorized<
+      T, detail::LaunchGatherKernelVectorized<is_axis_zero>::template Impl>(
+      MinAlignmentOf(params, out, slice_size), d, params, indices, out,
```

Comment:
This can be a bit confusing since the user will get vectorized operations or not depending on memory allocation patterns.  Do you have future plans to make this more robust around memory alignment (for instance by using non-vector loads/stores for the unaligned prefix)?  Or do you think it doesn't matter in practice?

Reply:
I believe the allocator always returns aligned pointers, so I don't expect the alignment to vary.
In addition, doing non-vectorized ld/st for the prefix is not practical for multi-dimensional operations with a non-aligned stride because each row has a different alignment.

Reply:
Great, can you please add a comment that states this?

Reply:
Added.